File: 673/673.cpp
#include<iostream>
#include<stack>
#include<string>
using namespace std;

bool process(string line){
    stack<char> stk;

    for(string::iterator it = line.begin(); it != line.end(); it++){
        char c = *it;
        switch(c){
            case '(':
            case '[':
                stk.push(c);
                break;
            case ')':
            case ']':
                if ( stk.empty())
                    return false;
                {
                char last = stk.top();
                if (       (c == ')' and last != '(') 
                        or (c == ']' and last != '['))
                    return false;
                }
                stk.pop();
                break;
            default:
                cerr << "Invalid Character: " 
                    << c << " (" << int(c) << ")" <<endl;
        }
    }

    return stk.empty();
}

int main(){
    int n;
    cin >> n;
    cin.ignore(2, '\n');

    for(int i = 0; i < n; i++){
        string line;
        getline(cin, line);

        cout << (process(line)? "Yes": "No")
            << '\t' << line
            << endl;
    }

    return 0;
}

File: 673/673.py
#!/usr/bin/env python3

def run(s):
    stack = []
    for c in s:
        if c == '(' or c == '[':
            stack.append(c)
        elif c == ')' or c == ']':
            if not stack:
                return False
            last = stack.pop()
            if (       (c == ')' and last != '(') 
                    or (c == ']' and last != '[')):
                return False
        else:
            raise ValueError('invalid character in string: {} ({})'.format(
                c, ord(c)))
    return not stack


if __name__ == '__main__':
    import sys
    next(sys.stdin) # ignore first line
    for case in sys.stdin:
        case = case.strip()
        print('Yes' if run(case) else 'No', case, sep="\t")

